### DISCUSSION TITLE
Problem with Martensite Parent Grain Reconstruction

### DISCUSSION THREAD

**Discussion by arthurribe** [mtex-toolbox/mtex#973] [Ask Anything]
Hello Mtex comunity,

I'm trying to perform a martensite parent Grain reconstruction using the following code; The problem occurs when I try to compute parent grai orientation it gives the output

"Error using orientation/dot_outer (line 25)
comparing orientations of different phase not yet supported

Error in quaternion/angle_outer (line 15)
omega = real(2*acos(abs(dot_outer(q1,q2,varargin{:}))));

Error in calcParent (line 60)
  fit = angle_outer(mori,p2c,'noSym2');"

Do I have to use a older version of Mtex to get the result?

Code:

% extract fcc and bcc symmetries
csBCC = ebsd.CSList{2}; % austenite bcc:
>> csFCC = ebsd.CSList{2}; % martensite fcc:
 
>> csFCC = ebsd.CSList{2}; % martensite fcc:
>> % grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
>> % remove small grains
ebsd(grains(grains.grainSize < 4)) = [];
>> % reidentify grains with small grains removed:
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
>> % plot the data and the grain boundaries
plot(ebsd('Iron bcc'),ebsd('Iron bcc').orientations,'figSize','large')
hold on
plot(grains.boundary,'linewidth',2)
hold off

 
>> % initial gues is Kurdjumov Sachs
KS = orientation.KurdjumovSachs(csFCC,csBCC);
>> % get neighbouring grain pairs
grainPairs = grains.neighbors;
>> % ignore pairs with misorientation angle smaller then 5 degree
mori = grains(grainPairs).meanOrientation;
grainPairs(angle(mori(:,1),mori(:,2)) < 5*degree,:) = [];
>> % compute an optimal parent to child orientation relationship
[fcc2bcc, fit] = calcParent2Child(grains(grainPairs).meanOrientation,KS);
 
  
>> [gB,pairId] = grains.boundary.selectByGrainId(grainPairs);
>> plot(ebsd('Iron bcc'),ebsd('Iron bcc').orientations,'figSize','large')
hold on;
plot(gB,fit(pairId) ./ degree,'linewidth',3,'smooth')
hold off

>> omega = linspace(0,5)*degree;
threshold = 2*degree;
tol = 1.5*degree;
>> % compute the probabilities
prob = 1 - 0.5 * (1 + erf(2*(fit - threshold)./tol));
>> % the corresponding similarity matrix
A = sparse(grainPairs(:,1),grainPairs(:,2),prob,length(grains),length(grains));
>> p = 1.6; % inflation power:
A = mclComponents(A,p);
>> % merge grains according to the adjacency matrix A
[parentGrains, parentId] = merge(grains,A);
>> % ensure grainId in parentEBSD is set up correctly with parentGrains
parentEBSD = ebsd;
parentEBSD('indexed').grainId = parentId(ebsd('indexed').grainId);
>> plot(ebsd('Iron bcc'),ebsd('Iron bcc').orientations,'figSize','large')
hold on;
plot(parentGrains.boundary,'linewidth',4)
hold off
 
>> % the measured child orientations
childOri = grains('Iron bcc').meanOrientation;
>> % the parent orientation we are going to compute
parentOri = orientation.nan(max(parentId),1,fcc2bcc.CS);
>> fit = inf(size(parentOri));
>> weights = grains('Iron bcc').grainSize;
>> % loop through all parent grains
for k = 1:max(parentId)
  if nnz(parentId==k) > 1
    % compute the parent orientation from the child orientations
    [parentOri(k),fit(k)] = calcParent(childOri(parentId==k), fcc2bcc,'weights',weights((parentId==k)));
  end
  progress(k,max(parentId));
end

>> % update mean orientation of the parent grains
parentGrains(fit<5*degree).meanOrientation = parentOri(fit<5*degree);
parentGrains = parentGrains.update;
>> % merge grains with similar orientation
[parentGrains, mergeId] = merge(parentGrains,'threshold',3*degree);
parentEBSD('indexed').grainId = mergeId(parentEBSD('indexed').grainId);
>> plot(parentGrains('Iron fcc'),parentGrains('Iron fcc').meanOrientation)
 
Warning: There is no such phase "Iron fcc". Maybe you mispelled it? 
 
Warning: There is no such phase "Iron fcc". Maybe you mispelled it? 
>> plot(parentGrains('Iron bcc'),parentGrains('Iron bcc').meanOrientation)

>> % compute variantId and packetId
[variantId,packetId] = calcVariantId(parentOri(parentId),childOri,fcc2bcc);
Warning: Packet ID calculation assuming {111}_p||{110}_c 
> In calcVariantId (line 65) 
>> % associate to each packet id a color and plot
color = ind2color(packetId);
plot(grains,color)
>> hold on
plot(parentGrains.boundary,'linewidth',1)
>> % outline a specific parent grain
hold on
id = parentGrains.findByLocation([100,80]);
plot(parentGrains(id).boundary,'linewidth',3,'lineColor','w')
hold off
>> % the measured child orientations that belong to parent grain 279
childOri = ebsd(parentEBSD.grainId==id).orientations;
plotPDF(childOri,Miller(0,0,1,csBCC),'MarkerSize',3)
  I'm plotting 208 random orientations out of 380 given orientations
  You can specify the the number points by the option "points".
  The option "all" ensures that all data are plotted
>> 
% the orientation of parent grain 279
hold on
parentOri = parentGrains(id).meanOrientation;
plot(parentOri.symmetrise * Miller(0,0,1,csFCC))
>> % the theoretical child variants
childVariants = variants(fcc2bcc, parentOri);
plotPDF(childVariants, 'markerFaceColor','none','linewidth',2,'markerEdgeColor','orange')
hold off
>> % consider only martensite pixels that now belong to austenite grains
isNowFCC = parentGrains.phaseId(max(1,parentEBSD.grainId)) == 3 & parentEBSD.phaseId == 2;
>> % compute parent orientation
[parentEBSD(isNowFCC).orientations, fit] = calcParent(ebsd(isNowFCC).orientations,...
  parentGrains(parentEBSD(isNowFCC).grainId).meanOrientation,fcc2bcc);
Error using orientation/dot_outer (line 25)
comparing orientations of different phase not yet supported

Error in quaternion/angle_outer (line 15)
omega = real(2*acos(abs(dot_outer(q1,q2,varargin{:}))));

Error in calcParent (line 60)
  fit = angle_outer(mori,p2c,'noSym2');

**Comment by ralfHielscher** [comment]
Hi Arthur,

we would strongly recommend using the parentGrainReconstructor class included in the most recent MTEX release. 
It is documented here: https://mtex-toolbox.github.io/MaParentGrainReconstruction.html
It is much easier to use. In your case there seems to be a mistake with the phase names. Though there is too much code to go through it.

Ralf.

**Reply by arthurribe** [reply]
Thanks for the quick response, Ralf.
When I perform the code you suggested I get the following error  for this job

>> % initial guess for the parent to child orientation relationship
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild)
Error using Miller (line 87)
No crystal symmetry has been specified when defining a crystal direction!

Error in orientation.KurdjumovSachs (line 238)
    mori = orientation.map(Miller(1,1,1,csGamma),Miller(0,1,1,csAlpha),...
